IDFSP-5663: Live reload controller changes

AalokAhluwalia opened the pull request:
IDFSP-5663

AalokAhluwalia committed:
IDFSP-5663: Live reload controller changes

AalokAhluwalia committed:
Resolve merge conflict

dfarr commented on controllers/sensor/resource.go:
```diff
@@ -66,6 +67,18 @@ func Reconcile(client client.Client, eventBus *eventbusv1alpha1.EventBus, args *
 
+	if args.Sensor.Spec.LiveReload {
+
+		sensorConfigmMapName := fmt.Sprintf("sensor-cm-%s", args.Sensor.Name)
```
I'm wondering if we should remove `-cm` from the name since it's redundant?

dfarr commented on sensors/cmd/start.go:
```diff
@@ -4,19 +4,19 @@ import (
 	"encoding/base64"
```
Can we remove the changes to this file?

dfarr commented on pkg/apis/sensor/v1alpha1/types.go:
```diff
@@ -113,6 +113,9 @@ type SensorSpec struct {
 	// LoggingFields add additional key-value pairs when logging happens
 	// +optional
 	LoggingFields map[string]string `json:"loggingFields" protobuf:"bytes,8,rep,name=loggingFields"`
+	// LiveReload defines whether sensor pods should restart on changes to the deployment spec.
```
This is a tricky line to write since pods are somewhat immutable (although I think some fields can be updated such as annotations) so they are never restart per-se, instead they are deleted+created. Also some of the fields in the sensor will still result in "restarts" even after this change.

How about
```
LiveReload propagates sensor changes to the sensor deployment without the need for recreating pods, defaults to false
```

dfarr commented on controllers/sensor/resource.go:
```diff
@@ -66,6 +67,18 @@ func Reconcile(client client.Client, eventBus *eventbusv1alpha1.EventBus, args *
+			return err
+		}
+		sensorContent := map[string]string{"sensor.yaml": string(serializedBytes)}
+		updateOrCreateConfigMap(ctx, client, sensorConfigmMapName, sensor.Namespace, sensorContent)
```
I'm wondering if we should call this `buildConfigMap` and return the cm object in order to match the `buildDeployment` pattern?

dfarr commented on controllers/sensor/resource.go:
```diff
@@ -212,12 +227,24 @@ func buildDeployment(args *AdaptorArgs, eventBus *eventbusv1alpha1.EventBus) (*a
+					Name: fmt.Sprintf("sensor-cm-%s", args.Sensor.Name),
+				}}},
+		})
+		volumeMounts = append(volumeMounts, corev1.VolumeMount{Name: "sensor-config-volume", MountPath: "/sensor-definition"})
```
Mind if we set mountpath to `/etc/sensor` to match this [eventbus mount path](https://github.com/intuit-data-os/argo-events/blob/master/common/common.go#L47)? Also, if we use a constant in `common.go` we could use that constant to read from the same file in the sensor.

dfarr reviewed (changes requested):
I'm wondering if it would be better to use json as opposed to yaml, but I can't really think of any good arguments one way or the other. Can you think of any reason we should prefer one over the other?

AalokAhluwalia committed:
Address pr fixes ex linting

AalokAhluwalia committed:
lint and codegen

AalokAhluwalia committed:
fix test

AalokAhluwalia committed:
Env var for mount path and test change

AalokAhluwalia committed:
lint

AalokAhluwalia committed:
Refactor to use json instead of yaml

dfarr committed:
Live reload sensor (#29)

When an environment variable is present we switch to live reload mode. The file path specified by the env variable is watched for changes. When a change occurs, the current context is cancelled, causing the sensor to stop processing messages and restarted with a new context.

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Refactor common constants for live reload

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add owner reference to configmap for cleanup

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Update comments

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Add live reload functional test

Signed-off-by: David Farr <david_farr@intuit.com>

dfarr committed:
Merge branch 'master' into IDFSP-5663

Signed-off-by: David Farr <david_farr@intuit.com>
